pkg/p2p/peer/processing/chainsync/sync.go

```diff
@@ -87,13 +87,18 @@ func (s *ChainSynchronizer) Synchronize(blkBuf *bytes.Buffer, peerInfo string) e
 		return nil
 	}
 
+	// Does the block come directly after our most recent one?
 	if diff == 1 {
-		// Write bufio.Reader into a bytes.Buffer so we can send it over the event bus.
+		// Write bufio.Reader into a bytes.Buffer and unmarshal it so we can send it over the event bus.
 		buf := new(bytes.Buffer)
 		if _, err := buf.ReadFrom(r); err != nil {
 			return err
 		}
 
+		if err := message.UnmarshalBlock(buf, blk); err != nil {
+			return err
+		}
+
 		msg := message.New(topics.Block, buf)
 		s.publisher.Publish(topics.Block, msg)
```

Comment:
That's a good catch.

Would it make sense (for sake of clarity) to have `topics.BlockWire` (accepting bytes.Buffer) that comes from the wire and `topics.Block` that is internally propagated only (accepting block.Block{} )?

 If we apply such naming convention (any messages on the wire *Wire suffixed ) we could avoid facing similar issues in future.

Reply:
Actually, why don't we gossip buffer instead of block.Block{}?

```
func (c *Chain) propagateBlock(blk block.Block) error {
	buffer := topics.Block.ToBuffer()
	if err := message.MarshalBlock(&buffer, &blk); err != nil {
		return err
	}

	msg := message.New(topics.Block, blk)
	c.eventBus.Publish(topics.Gossip, msg)
	return nil
}
```

Reply:
This function never seems to be called - I think it might be legacy code from when we did not use the `Inv` messages yet.

As for `topics.BlockWire`, I don't think we need it. Wire encoded blocks are passed directly to the `Synchronizer`, who will now unmarshal it and then publish it on `topics.Block`. So no such topic would be needed.

Reply:
It's good we have now message payload of type` interface{}`. My suggestion is to make it easier for anyone who publishes or collects a message to know if it is a buffer or generic type. This could be achieved with naming convention or at least some comments in topics package.

Currently it's very unclear what payload each topic pairs with. That might lead to more panic issues like this one.